server/src/mahjong/meld.rs

```diff
@@ -0,0 +1,175 @@
+            return Err(IllegalChiError::NotASequenceError(x, y, z));
+        }
+
+        Ok(Self(x, y, z))
+    }
+
+    pub fn get_0(&self) -> &Tile {
+        &self.0
+    }
+
+    pub fn get_1(&self) -> &Tile {
+        &self.1
+    }
+
+    pub fn get_2(&self) -> &Tile {
+        &self.2
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Pong(Tile);
+
+impl Pong {
+    pub fn new(tile: Tile) -> Result<Self, IllegalPongError> {
+        match is_playable_tile(&tile) {
+            true => Ok(Self(tile)),
+            false => Err(IllegalPongError::IllegalTileError(tile)),
+        }
+    }
+
+    pub fn get_0(&self) -> &Tile {
+        &self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Gang(Tile);
+
+impl Gang {
+    pub fn new(tile: Tile) -> Result<Self, IllegalGangError> {
+        match is_playable_tile(&tile) {
+            true => Ok(Self(tile)),
+            false => Err(IllegalGangError::IllegalTileError(tile)),
+        }
+    }
+    pub fn get_0(&self) -> &Tile {
+        &self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct AnGang(Tile);
+
+impl AnGang {
+    pub fn new(tile: Tile) -> Result<Self, IllegalAnGangError> {
+        match is_playable_tile(&tile) {
+            true => Ok(Self(tile)),
+            false => Err(IllegalAnGangError::IllegalTileError(tile)),
+        }
+    }
+    pub fn get_0(&self) -> &Tile {
+        &self.0
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum IllegalChiError {
+    #[error("Illegal to chi tile: {0:?}. Chi can only be performed on Wan, Suo and Tong")]
+    IllegalTileError(Tile),
+    #[error("Tiles {0:?}, {1:?}, {2:?} are not a valid sequence")]
+    NotASequenceError(Tile, Tile, Tile),
+    #[error("Tiles {0:?}, {1:?}, {2:?} are not the same suit")]
+    NotSameSuitError(Tile, Tile, Tile),
+}
+
+#[derive(Debug, Error)]
+pub enum IllegalPongError {
+    #[error(
+        "Illegal to pong tile: {0:?}. Pong can only be performed on Wan, Suo, Tong, Feng, Dragon"
+    )]
+    IllegalTileError(Tile),
+}
+
+#[derive(Debug, Error)]
+pub enum IllegalGangError {
+    #[error(
+        "Illegal to gang tile: {0:?}. Gang can only be performed on Wan, Suo, Tong, Feng, Dragon"
+    )]
+    IllegalTileError(Tile),
+}
+
+#[derive(Debug, Error)]
+pub enum IllegalAnGangError {
+    #[error("Illegal to angang tile: {0:?}. An gang can only be performed on Wan, Suo, Tong, Feng, Dragon")]
+    IllegalTileError(Tile),
+}
+
+#[cfg(test)]
+mod tests {
+    use claims::{assert_err, assert_ok};
+
+    use crate::mahjong::{
+        meld::{AnGang, Chi, Gang, Pong},
+        tile::{AnimalType, HuaType, Tile},
+    };
+
+    #[test]
+    fn test_create_chi() {
+        assert_err!(Chi::new(
+            Tile::Wan(1),
+            Tile::Wan(2),
+            Tile::Animal(AnimalType::Cat)
+        ));
+        assert_err!(Chi::new(Tile::Wan(1), Tile::Wan(2), Tile::Tong(3)));
+        assert_err!(Chi::new(Tile::Wan(1), Tile::Wan(2), Tile::Wan(4)));
```

Comment:
Would be good to assert on the expected error message. This applies to the other tests that do assert_err! as well

Reply:
Will do in the future, display not properly implemented for the errors yet.